### PR TITLE
alert: Add CWE Alert Tag when building and CWE ID has been set

### PR DESCRIPTION
**Pull request by kingthorin**
I also looked at doing this for WASC IDs but they don't have a handy URL, we'd have to add a pre-populated map which isn't impossible, but did seem like too much work for this morning 😀

### CODE REVIEW

**Review by ricekot — COMMENTED** [review]
The function `Alert#setCweId` should be updated too?
https://github.com/zaproxy/zaproxy/blob/24a4eb5afa852605d2bba7b4f9e7bb9c305667ef/zap/src/main/java/org/parosproxy/paros/core/scanner/Alert.java#L931

**Comment by kingthorin** [comment]
I dunno we didn't update the old alert functionality when we implemented addTag/removeTag.

I'm happy to do either, but was trying to stay consistent.

**Comment by thc202** [comment]
Whatever we do we should document these side effects to avoid surprises (e.g. I can see this breaking our unit tests as there will be more tags now).

File: zap/src/main/java/org/parosproxy/paros/core/scanner/Alert.java [hunk 1]
```diff
@@ -193,6 +195,10 @@ public static Source getSource(int id) {
         "False Positive", "Low", "Medium", "High", "Confirmed"
     };
 
+    private static final String CWE_TEMPLATE_KEY = "CWE-{0}";
+    private static final String CWE_TEMPLATE_URL =
+            "https://cwe.mitre.org/data/definitions/{0}.html";
```
Comment (thc202) [inline zap/src/main/java/org/parosproxy/paros/core/scanner/Alert.java:200]: It would be simpler to just concatenate. Having hardcoded link in core :/ but guess there's no way around (it other than allowing to set from add-ons).

File: zap/src/main/java/org/parosproxy/paros/core/scanner/Alert.java [hunk 2]
```diff
@@ -1142,6 +1148,7 @@ public Builder setInputVector(String inputVector) {
 
         public Builder setCweId(int cweId) {
             this.cweId = cweId;
+            this.addCweTag(cweId);
```
Comment (thc202) [inline zap/src/main/java/org/parosproxy/paros/core/scanner/Alert.java:1151]: - This should be done when the alert is being built. There should be a test showing that setting multiple IDs does not lead to multiple tags.
- This behaviour should be documented (in `build`?).

**Comment by kingthorin** [comment]
Okay how about this:

1. I'll deprecate the existing setCwe(int) and add setCwe(int, boolean). This can easily be handled with a regex search and replace (with grouping/interpolation) when we update to new core. Where boolean true is to add the tag.
2. In Builder.build() when boolean true then it's added.
3. Add functionality/tests to not set multiple tags.
4. In the Alert functionality I'll add an addCweTag() method? Or also do 1-3 there?

**Comment by kingthorin** [comment]
Crew I need some feedback on this so that it can be moved ahead.

**Comment by kingthorin** [comment]
Anyone out there?

**Comment by ricekot** [comment]
I like the original approach of adding the tags in the builder, if a valid CWE ID is set.

**Comment by thc202** [comment]
I'm also happy with the builder way, my only concern is the URL (though that one should stay stable…).

**Comment by kingthorin** [comment]
Thank you both! I'll get this reworked.

File: zap/src/test/java/org/parosproxy/paros/core/scanner/AlertUnitTest.java [hunk 3]
```diff
@@ -157,4 +159,55 @@ void shouldBuildAlertWithNoTagsWhenOneRemoved() {
...
+    @Test
+    void shouldAddOneCweTagEvenIfCweIdSetTwice() {
+        // Given
+        int cwe = 618;
+        int cwe2 = 619;
+        String cweUrl = "https://cwe.mitre.org/data/definitions/619.html";
+        String cweKey = "CWE-" + cwe2;
+        Builder builder = new Alert.Builder();
+        builder.setCweId(cwe);
+        builder.setCweId(cwe2);
```
Comment (thc202) [inline zap/src/test/java/org/parosproxy/paros/core/scanner/AlertUnitTest.java:203]: Call build before this line as well (the tag needs to be added to the alert rather than the builder).

File: zap/src/main/java/org/parosproxy/paros/core/scanner/Alert.java [hunk 4]
```diff
@@ -1242,6 +1246,9 @@ public Builder removeTag(String tag, String value) {
          * <p>The alert URI defaults to the one from the {@code HistoryReference} or {@code
          * HttpMessage} if set.
          *
+         * <p><strong>Note:</strong> If the Alert has a CWE set then an associated Tag will be added
+         * during @link Builder#build()
```
Comment (thc202) [inline zap/src/main/java/org/parosproxy/paros/core/scanner/Alert.java:1250]: Missing {}.

File: zap/src/main/java/org/parosproxy/paros/core/scanner/Alert.java [hunk 5]
```diff
@@ -1277,10 +1284,26 @@ public final Alert build() {
             if (alertRef != null) {
                 alert.setAlertRef(alertRef);
             }
+            addCweTag(cweId);
             alert.setTags(tags);
```
Comment (thc202) [inline zap/src/main/java/org/parosproxy/paros/core/scanner/Alert.java:1288]: Not part of this PR but we should be setting a copy, otherwise changes done afterwards to the builder might affect an alert already built.

Comment (kingthorin) [reply zap/src/main/java/org/parosproxy/paros/core/scanner/Alert.java:1288]: I looked at this quickly, from my understanding it would need a deep copy which there doesn't seem to be a simple solution to. https://www.baeldung.com/java-copy-hashmap

I only checked that one article so I could have just missed the point.

**Comment by kingthorin** [comment]
Tweaked

File: zap/src/test/java/org/parosproxy/paros/core/scanner/AlertUnitTest.java (hunk 3 above)
Comment (thc202) [reply zap/src/test/java/org/parosproxy/paros/core/scanner/AlertUnitTest.java]: The semantics of the (new) test were changed, the second call to `setCweId` should still be done to the builder.

File: zap/src/main/java/org/parosproxy/paros/core/scanner/Alert.java (hunk 5 above)
Comment (thc202) [reply zap/src/main/java/org/parosproxy/paros/core/scanner/Alert.java:1288]: e.g. `new HashMap(tags)` would be enough to copy. Now that I'm reading the code again, it will have to be addressed in this PR anyway, to add the CWE tag.

File: zap/src/test/java/org/parosproxy/paros/core/scanner/AlertUnitTest.java (hunk 3 above)
Comment (kingthorin) [reply zap/src/test/java/org/parosproxy/paros/core/scanner/AlertUnitTest.java]: Ok, I misinterpreted something said earlier. Will tweak.

File: zap/src/main/java/org/parosproxy/paros/core/scanner/Alert.java (hunk 5 above)
Comment (kingthorin) [reply zap/src/main/java/org/parosproxy/paros/core/scanner/Alert.java:1288]: Okay.

**Comment by kingthorin** [comment]
Okay think I got this into a more expected state.

File: zap/src/main/java/org/parosproxy/paros/core/scanner/Alert.java [hunk 6]
```diff
@@ -1277,10 +1284,32 @@ public final Alert build() {
             if (alertRef != null) {
                 alert.setAlertRef(alertRef);
             }
-            alert.setTags(tags);
+            addCweTag(cweId);
+            alert.setTags(new HashMap<>(tags == null ? new HashMap<>() : tags));
```
Comment (thc202) [inline zap/src/main/java/org/parosproxy/paros/core/scanner/Alert.java:1288]:
```suggestion
            alert.setTags(withCweTag(tags, cweId));
```
and let `withCweTag` add the tag to the new map if/when needed, it should be simpler than add the tag to the `tags` and then having to undo.

**Comment by kingthorin** [comment]
Tweaked again.

**Comment by kingthorin** [comment]
Okay hopefully this is finally good to go.

**Comment by kingthorin** [comment]
I had to re-work it a bit more because I was getting NPE and unmodifiable map exceptions while working on something else this evening. Hopefully it's in a good state now.

**Review by thc202 — APPROVED** [review]

**Comment by thc202** [comment]
Thank you!

**Comment by kingthorin** [comment]
For the record this will mean that we need to update "expected mappings" type unit tests when zap-extensions migrates to 2.15.

**Review by psiinon — APPROVED** [review]
